File: app/src/main/java/com/jbrunton/mymovies/movies/BaseMovieViewStateFactory.java
package com.jbrunton.mymovies.movies;

import com.google.common.base.Optional;
import com.jbrunton.entities.Movie;
import com.jbrunton.mymovies.shared.LoadingViewStateFactory;

import org.joda.time.LocalDate;

public class BaseMovieViewStateFactory {
    protected final LoadingViewStateFactory loadingViewStateFactory = new LoadingViewStateFactory();

    private String emptyIfNull(Optional<String> s) {
        if (s.isPresent()) {
            return s.get();
        } else {
            return "";
        }
    }

    private String convertReleaseDate(Optional<LocalDate> date) {
        if (date.isPresent()) {
            return Integer.toString(date.get().getYear());
        } else {
            return "";
        }
    }

    protected <T extends BaseMovieViewState.Builder<T>> T setDefaults(T builder, Movie movie) {
        return builder.movieId(movie.id())
                .title(movie.title())
                .yearReleased(convertReleaseDate(movie.releaseDate()))
                .rating("&#9734; " + movie.rating())
                .posterUrl(emptyIfNull(movie.posterUrl()))
                .backdropUrl(emptyIfNull(movie.backdropUrl()));
    }
}

File: app/src/main/java/com/jbrunton/mymovies/movies/BaseMovieViewStateFactory.kt
package com.jbrunton.mymovies.movies

import com.google.common.base.Optional
import com.jbrunton.entities.Movie
import com.jbrunton.mymovies.shared.LoadingViewStateFactory
import org.joda.time.LocalDate

open class BaseMovieViewStateFactory {
    protected val loadingViewStateFactory = LoadingViewStateFactory()

    protected fun <T : BaseMovieViewState.Builder> setDefaults(builder: T, movie: Movie): T {
        builder.movieId = movie.id
        builder.title = movie.title
        builder.yearReleased = convertReleaseDate(movie.releaseDate)
        builder.rating = "&#9734; " + movie.rating
        builder.posterUrl = emptyIfNull(movie.posterUrl)
        builder.backdropUrl = emptyIfNull(movie.backdropUrl)
        return builder
    }

    private fun emptyIfNull(s: Optional<String>): String {
        return if (s.isPresent) {
            s.get()
        } else {
            ""
        }
    }

    private fun convertReleaseDate(date: Optional<LocalDate>): String {
        return if (date.isPresent) {
            Integer.toString(date.get().year)
        } else {
            ""
        }
    }

}
